src/operations/reweight.lisp

```diff
@@ -83,35 +84,42 @@
                             `(BROADCAST-LOCK ,targets)
                             `(CHECK-ROOTS (,source-root))
                             `(BROADCAST-PINGABILITY ,targets :SOFT)
-                            `(CHECK-REWEIGHT ,pong)
-                            ;; it bugs me a bit having to switch this back & forth
+                            `(CHECK-REWEIGHT (,source-root) ,pong)
                             `(BROADCAST-PINGABILITY ,targets :NONE)
                             `(BROADCAST-REWEIGHT (,source-root) ,weight)
                             `(BROADCAST-PINGABILITY ,targets :SOFT)
                             `(CHECK-REWINDING (,source-root) ,pong 0)
                             `(BROADCAST-UNLOCK)
                             `(HALT)))))
 
-(define-process-upkeep ((supervisor supervisor)) (CHECK-REWEIGHT pong)
+(define-process-upkeep ((supervisor supervisor)) (CHECK-REWEIGHT roots original-pong)
   "Because `CHECK-PONG' doesn't do a global check, we potentially can end up with a reweighting when we shouldn't. This fixes that by making sure that there are no lower-weight recommendations available before we begin reweighting."
   (unless (process-lockable-aborting? supervisor)
-    (with-slots (source-root weight) pong
-      (let ((listen-channel (register)))
-        (send-message source-root (make-message-soft-scan
-                                   :reply-channel listen-channel
-                                   :local-root source-root
-                                   :weight 0
-                                   :repeat? t))
-        (sync-receive (listen-channel pong-message)
-          (message-pong
-           (unregister listen-channel)
-           (when (< (message-pong-weight pong-message) weight)
-             (setf (process-lockable-aborting? supervisor) t))))))))
+    (let ((check-pong nil)
+          (original-weight (message-pong-weight original-pong)))
+      (flet ((payload-constructor ()
+               (make-message-soft-scan :weight 0
+                                       :internal-roots roots
+                                       :strategy :STAY)))
+        (with-replies (replies
+                       :message-type message-pong
+                       :message-unpacker identity)
+                      (send-message-batch #'payload-constructor roots)
+          (loop :for reply :in replies :unless (null reply)
```

Comment:
not actually recommended, but: `(reduce (lambda (x y) (or (and x y (unify-pongs x y)) x y)) replies :start nil)` and then we could even drop the nil casing in `unify-pongs` itself